File: leo2-node/src/main/java/org/deku/leo2/node/LocalStorage.java
package org.deku.leo2.node;

import sx.LazyInstance;

import java.io.File;

/**
 * Created by masc on 26.06.15.
 */
public class LocalStorage {
    private static LazyInstance<LocalStorage> mInstance = new LazyInstance(LocalStorage::new);

    // Directories
    private File mHomeDirectory;
    private File mDataDirectory;
    private File mConfigurationDirectory;

    // Files
    private File mApplicationConfigurationFile;
    private File mIdentityConfigurationFile;
    private File mLogFile;
    private File mActiveMqDataDirectory;

    /**
     * c'tor
     */
    private LocalStorage() {
        mHomeDirectory = new File(System.getProperty("user.home"), ".leo2");
        mDataDirectory = new File(this.getHomeDirectory(), "data");

        mApplicationConfigurationFile = new File(this.getHomeDirectory(), "leo2.properties");
        mIdentityConfigurationFile = new File(this.getDataDirectory(), "identity.properties");
        mLogFile = new File(this.getHomeDirectory(), "leo2.log");
        mActiveMqDataDirectory = new File(this.getDataDirectory(), "activemq");
    }

    /**
     * Initializes local storage, takes care directories exist etc.
     */
    public void initialize() {
        mHomeDirectory.mkdirs();
        mDataDirectory.mkdirs();
    }

    /** Singleton accessor */
    public static LocalStorage instance() {
        return mInstance.get();
    }

    /**
     * Local home directory
     * @return
     */
    public File getHomeDirectory() {
        return mHomeDirectory;
    }

    /**
     * Local data directory
     * @return
     */
    public File getDataDirectory() {
        return mDataDirectory;
    }

    /**
     * Local application configuration file
     * @return
     */
    public File getApplicationConfigurationFile() {
        return mApplicationConfigurationFile;
    }

    /**
     * Local identity configuration file
     * @return
     */
    public File getIdentityConfigurationFile() {
        return mIdentityConfigurationFile;
    }

    /**
     * Local log file
     * @return
     */
    public File getLogFile() {
        return mLogFile;
    }

    /**
     * Local embedded activemq data directory
     * @return
     */
    public File getActiveMqDataDirectory() {
        return mActiveMqDataDirectory;
    }
}

File: leo2-node/src/main/java/org/deku/leo2/node/LocalStorage.kt
package org.deku.leo2.node

import com.google.common.base.Strings
import org.apache.commons.lang3.SystemUtils
import org.apache.commons.logging.Log
import org.apache.commons.logging.LogFactory
import sx.LazyInstance

import java.io.File
import java.util.*
import kotlin.platform.platformStatic

/**
 * Local Storage
 * Created by masc on 26.06.15.
 */
class LocalStorage {
    private var log: Log = LogFactory.getLog(this.javaClass)

    companion object Singleton {
        private val instance: LocalStorage = LocalStorage()
        @platformStatic fun instance() : LocalStorage {
            return this.instance;
        }
    }

    /** c'tor */
    private constructor () {
        // Initialize directories
        var basePath: String
        var baseDirectory: String
        if (SystemUtils.IS_OS_WINDOWS) {
            basePath = System.getenv("ALLUSERSPROFILE")
            baseDirectory = "LeoZ"
        } else {
            basePath = System.getProperty("user.home")
            baseDirectory = ".leoZ"
        }
        if (Strings.isNullOrEmpty(basePath))
            throw UnsupportedOperationException("Basepath is empty");

        this.homeDirectory = File(basePath, baseDirectory)
        this.log.info("Home directory [${homeDirectory}]")
        this.logDirectory = File(this.homeDirectory, "log");
        this.dataDirectory = File(this.homeDirectory, "data")

        this.applicationConfigurationFile = File(this.homeDirectory, "leo2.properties")
        this.identityConfigurationFile = File(this.dataDirectory, "identity.properties")
        this.logFile = File(this.logDirectory, "leo2.log")
        this.activeMqDataDirectory = File(this.dataDirectory, "activemq")

        this.homeDirectory.mkdirs()
        this.dataDirectory.mkdirs()
        this.logDirectory.mkdirs()
    }

    // Directories
    /** Local home directory */
    val homeDirectory: File;
    /** Local data directory */
    val dataDirectory: File
    /** Local log directory */
    val logDirectory: File
    /** Local embedded activemq data directory */
    val activeMqDataDirectory: File

    // Files
    /** Local application configuration file */
    val applicationConfigurationFile: File
    /** Local identity configuration file */
    val identityConfigurationFile: File
    /** Local log file */
    val logFile: File
}
